libs/events/events_test.go

```diff
@@ -174,40 +172,38 @@ func TestAddAndRemoveListenerConcurrency(t *testing.T) {
 
 	evsw := NewEventSwitch()
 	err := evsw.Start()
-	if err != nil {
-		t.Errorf("Failed to start EventSwitch, error: %v", err)
-	}
+	require.NoError(t, err)
 	defer evsw.Stop()
 
 	done1 := make(chan struct{})
 	done2 := make(chan struct{})
 
+	// Must be executed concurrently to uncover the data race.
+	// 1. RemoveListener
 	go func() {
 		for i := 0; i < roundCount; i++ {
 			evsw.RemoveListener("listener")
 		}
-		done1 <- struct{}{}
+		close(done1)
 	}()
 
+	// 2. AddListenerForEvent
 	go func() {
 		for i := 0; i < roundCount; i++ {
-			index := i //it necessary for closure
+			index := i
 			evsw.AddListenerForEvent("listener", fmt.Sprintf("event%d", index),
 				func(data EventData) {
 					t.Errorf("should not run callback for %d.\n", index)
 					stopInputEvent = true
 				})
 		}
-		done2 <- struct{}{}
+		close(done2)
 	}()
 
 	<-done1
 	<-done2
 
-	close(done1)
-	close(done2)
-
-	evsw.RemoveListener("listener") // make sure remove last
+	evsw.RemoveListener("listener") // remove the last listener
```

Comment:
@goolAdapter why do we treat last listener specially?

Reply:
because AddListenerForEvent routine may finish after RemoveListener routine，at this time， the evsw have some "listener" callback as except.
our test is to demonstration RemoveListener can't clear up the cb because race time window.